### PR TITLE
Flow

### PR DESCRIPTION
**Pull request by zyndaras**

### CODE REVIEW

**Comment by zyndaras** [comment]
@daves125125

File: src/main/java/com/opinari/otf/filter/DefaultUrlIntegrityFilter.java [hunk 1]
```diff
@@ -12,7 +17,29 @@
...
+        return urlValidator.isValid(url);
+    }
+
+    private boolean checkUrlIsReachable(String urlString) {
+        try {
+            URL url = new URL(urlString);
+            HttpURLConnection.setFollowRedirects(false);
+            HttpURLConnection httpURLConnection = (HttpURLConnection) url.openConnection();
+            httpURLConnection.setRequestMethod("HEAD");
+            return (httpURLConnection.getResponseCode() == HttpURLConnection.HTTP_OK);
```
Comment (daves125125) [inline src/main/java/com/opinari/otf/filter/DefaultUrlIntegrityFilter.java]: Have a single return statement, declare the boolean at the start of the method

File: src/main/resources/com/opinari/otf/integration.xml [hunk 2]
```diff
@@ -13,9 +13,7 @@
 		http://www.springframework.org/schema/util http://www.springframework.org/schema/util/spring-util.xsd
 		http://www.springframework.org/schema/mvc http://www.springframework.org/schema/mvc/spring-mvc.xsd">
```
Comment (daves125125) [inline src/main/resources/com/opinari/otf/integration.xml]: Do you need MVC and util schema declarations?

File: src/main/resources/com/opinari/otf/integration.xml [hunk 3]
```diff
@@ -35,12 +33,25 @@
...
+                  output-channel="chainInputChannel"/>
+
+    <int:chain input-channel="chainInputChannel" output-channel="chainOutputChannel">
+        <int:filter id="urlIntegrityFilter" discard-channel="discardChannel">
+            <bean class="com.opinari.otf.filter.DefaultUrlIntegrityFilter"/>
+        </int:filter>
+        <int:filter id="blacklistFilter" discard-channel="discardChannel">
+            <bean class="com.opinari.otf.filter.DefaultBlacklistFilter"/>
+        </int:filter>
+        <int:service-activator>
```
Comment (daves125125) [inline src/main/resources/com/opinari/otf/integration.xml]: Chain looks good, but should we include the service activator in here too? 

Need to think about where the line between this chain and the next point in the integration starts / ends

File: src/main/java/com/opinari/otf/filter/DefaultBlacklistFilter.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package com.opinari.otf.filter;
+
+import com.opinari.otf.domain.Pin;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author Ivan Jimenez Diez
+ */
+@Component
```
Comment (daves125125) [inline src/main/java/com/opinari/otf/filter/DefaultBlacklistFilter.java]: we dont need any annotations yet, we're using the xml to reference the beans by fully qualifed package name / class

File: src/main/java/com/opinari/otf/service/DefaultRejectionService.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package com.opinari.otf.service;
+
+import com.opinari.otf.domain.Pin;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author Ivan Jimenez Diez
+ */
+@Component
```
Comment (daves125125) [inline src/main/java/com/opinari/otf/service/DefaultRejectionService.java]: we dont need any annotations yet, we're using the xml to reference the beans by fully qualifed package name / class

File: src/main/java/com/opinari/otf/filter/DefaultUrlIntegrityFilter.java [hunk 6]
```diff
@@ -12,7 +17,29 @@
...
+    private boolean validateUrlFormat(String url) {
+        String[] schemes = {"http", "https"};
+        UrlValidator urlValidator = new UrlValidator(schemes);
+        return urlValidator.isValid(url);
+    }
+
+    private boolean checkUrlIsReachable(String urlString) {
+        try {
+            URL url = new URL(urlString);
+            HttpURLConnection.setFollowRedirects(false);
```
Comment (daves125125) [inline src/main/java/com/opinari/otf/filter/DefaultUrlIntegrityFilter.java]: why follow redirects?

File: src/main/java/com/opinari/otf/filter/DefaultUrlIntegrityFilter.java [hunk 7]
```diff
@@ -12,7 +17,29 @@
...
+            HttpURLConnection httpURLConnection = (HttpURLConnection) url.openConnection();
+            httpURLConnection.setRequestMethod("HEAD");
```
Comment (daves125125) [inline src/main/java/com/opinari/otf/filter/DefaultUrlIntegrityFilter.java]: why HEAD?

File: src/main/resources/com/opinari/otf/integration.xml [hunk 8]
```diff
@@ -35,12 +33,25 @@
...
-                  output-channel="splitPinsOutputChannel" />
-
-    <int:filter id="urlIntegrityFilter" ref="defaultUrlIntegrityFilter"
-                input-channel="splitPinsOutputChannel" output-channel="filterOutputChannel"/>
-
-    <int:channel id="filterOutputChannel">
+                  output-channel="chainInputChannel"/>
+
+    <int:chain input-channel="chainInputChannel" output-channel="chainOutputChannel">
+        <int:filter id="urlIntegrityFilter" discard-channel="discardChannel">
```
Comment (daves125125) [inline src/main/resources/com/opinari/otf/integration.xml]: This is good. 

Thinking about the behaviour in the filters though, some of the logic may be time consuming (relatively speaking), like looking up URLs etc. 

I wonder if theres a way to call these filters in a sensible order, use async model to continue while waiting on resources or calling in parallel? Not a biggie but interesting to consider

File: src/main/java/com/opinari/otf/filter/DefaultUrlIntegrityFilter.java [hunk 9]
```diff
@@ -12,7 +17,29 @@
...
     public boolean isUrlValid(Pin pin) {
-        return true;
+        boolean isUrlWellFormed = validateUrlFormat(pin.getUrl());
+        boolean isUrlUp = checkUrlIsReachable(pin.getUrl());
+
+        return isUrlWellFormed && isUrlUp;
+    }
+
+    private boolean validateUrlFormat(String url) {
+        String[] schemes = {"http", "https"};
```
Comment (daves125125) [inline src/main/java/com/opinari/otf/filter/DefaultUrlIntegrityFilter.java]: make private static final

**Pull request merged**